src/Centipede/Filter/UrlFilter.php

```diff
@@ -4,25 +4,43 @@
 
 class UrlFilter implements FilterInterface
 {
-    private $baseUrl;
-
-    public function __construct($baseUrl)
+    public function filter($url, $previousUrl = null)
     {
-        $this->baseUrl = rtrim($baseUrl, '/');
-    }
+        // These are all the settings we'll need to recreate the url.
+        // They're also the different keys that parse_url can return.
+        $parseUrlSettings = array('scheme' => '', 'host' => '', 'path' => '', 'port' => '', 'query' => '');
 
-    public function filter($value)
-    {
-        $value = rtrim($value, '/');
+        $parseValueUrl = array_merge($parseUrlSettings, parse_url($url));
 
-        if (false !== $position = strpos($value, '#')) {
-            $value = substr($value, 0, $position);
+        if (strpos($url, '#') === 0) {
+            return null;
         }
 
-        if (null !== parse_url($value, PHP_URL_SCHEME)) {
-            return $value;
+        // Return the value if we already have an absolute URL
+        if (isset($parseValueUrl['scheme'])) {
+            return $parseValueUrl['scheme'].'://'.$parseValueUrl['host'].$parseValueUrl['port'].$parseValueUrl['path'].($parseValueUrl['query'] ? ('?'.$parseValueUrl['query']) : '');
```

Comment:
I'd extract this line and https://github.com/umpirsky/centipede-crawler/pull/8/files#diff-7d4372ca01ff1148f4d1b3d4b574354fR43 to separate function.